README.md
# `carpenterd-api-client`

[![Version npm](https://img.shields.io/npm/v/carpenterd-api-client.svg?style=flat-square)](https://www.npmjs.com/package/carpenterd-api-client)
[![License](https://img.shields.io/npm/l/carpenterd-api-client.svg?style=flat-square)](https://github.com/warehouseai/carpenterd-api-client/blob/master/LICENSE)
[![npm Downloads](https://img.shields.io/npm/dm/carpenterd-api-client.svg?style=flat-square)](https://npmcharts.com/compare/carpenterd-api-client?minimal=true)
[![Dependencies](https://img.shields.io/david/warehouseai/carpenterd-api-client.svg?style=flat-square)](https://github.com/warehouseai/carpenterd-api-client/blob/master/package.json)

The `carpenterd-api-client` is an API client for the [`carpenterd`][carpenterd]
build service.

## Install

Install `carpenterd-api-client` from the npm registry:
```
npm install --save carpenterd-api-client
```

## Usage

In all examples we assume that you've already initialized the client as
followed:

```js
'use strict';

const Carpenter = require('carpenterd-api-client');

const carpenter = new Carpenter('url-to-the-service');
```

As you can see in the example above, the `Carpenter` constructor requires one
argument:

- The URL of the carpenter API where we should send the requests to.

## API

### build

Trigger a new build on carpenter service. The data provided should have
the same structure and signature as `npm publish` posted JSON.

```js
carpenter.build({ data: {
  "name": "tester",               // name of the package
  "dist-tags": {
    "latest": "1.0.0"
  },
  "versions": {
    "1.0.0": {
      "name": "tester",
      "version": "1.0.0"
      ...
    }
  },
  "_attachment": ""               // base64 encoded binary blob
}}, function () {

});
```

### cancel

Cancel a build on carpenter service.

```js
carpenter.cancel({
  "pkg": "tester",                  // name of the package
  "version": "1.0.0",               // valid semver
  "env": "prod"                     // optional environment parameter
}}, function () {

});
```

## Tests

```sh
npm test
```

[carpenterd]: https://github.com/godaddy/carpenterd


CHANGELOG.md
# Changelog

- [#4] Implement retry logic for build and cancel method.
- [#3] Add collected documentation and badges.

[#3]: https://github.com/warehouseai/carpenterd-api-client/pull/3
[#4]: https://github.com/warehouseai/carpenterd-api-client/pull/4


test.js
var assume = require('assume');
var nock = require('nock');
var url = require('url');
var Carpenter = require('./');

describe('carpenter-api-client', function () {
  this.timeout(5e4);
  var carpenter, uri;

  beforeEach(function each() {
    uri = 'http://localhost:8999/';
    carpenter = new Carpenter(uri);
  });

  afterEach(function each() {
    carpenter = null;
  });

  it('can be configured with object', function () {
    carpenter = new Carpenter(url.parse(uri));

    assume(carpenter.base).to.be.an('string');
    assume(carpenter.base).to.equal(uri);
  });

  it('can be configured with an agent and url', function () {
    carpenter = new Carpenter({
      url: uri,
      agent: require('http').Agent
    });

    assume(carpenter.agent);
  });

  it('can be configured with default timeout and defaults to 0', function () {
    assume(carpenter).to.have.property('timeout', 0);

    carpenter = new Carpenter({
      url: uri,
      timeout: 2 * 60 * 1000
    });

    assume(carpenter).to.have.property('timeout', 120000);
  });

  it('can be configured with retry and defaults to { retries: 5, min: 500, max: 10000 }', function () {
    assume(carpenter.retryOpts).eqls({ retries: 5, min: 500, max: 10000 });
    carpenter = new Carpenter({
      url: uri,
      retry: {
        retries: 3,
        min: 100,
        max: 500
      }
    });

    assume(carpenter.retryOpts).eqls({ retries: 3, min: 100, max: 500 });
  });

  it('can be configured with an agent and url object', function () {
    carpenter = new Carpenter({
      url: url.format(uri),
      agent: require('http').Agent
    });

    assume(carpenter.base).to.equal(uri);
    assume(carpenter.agent);
  });

  it('can be configured with an api version', function () {
    [1, '1', 'v1'].forEach(function (version) {
      carpenter = new Carpenter({ uri, version });

      assume(carpenter.version).equals('');
    });
  });

  it('defaults to the v2 api', function () {
    assume(new Carpenter({ uri }).version).equals('v2');
    assume(new Carpenter({ uri, version: 'not one' }).version).equals('v2');
  });

  describe('#build', function () {
    var options = {
      data: {
        data: { name: 'foo-bar' },
        promote: true
      }
    };

    it('sends a request to /v2/build', function (next) {
      next = assume.wait(2, next);

      nock(uri)
        .post('/v2/build')
        .reply(200, function reply(uri, body) {
          assume(body.data.name).equals('foo-bar');
          assume(body.promote).equals(true);
          nock.cleanAll();
          next();

          return {};
        });

      carpenter.build(options, next);
    });

    it('sends a request to /v2/build with retries and fails after retries', next => {
      carpenter = new Carpenter({
        url: uri,
        retry: {
          retries: 3,
          min: 100,
          max: 500
        }
      });
      nock(uri, {
        badheaders: ['whatever']
      })
        .post('/v2/build')
        .reply(200);

      carpenter.build(Object.assign({}, options, {
        headers: {
          whatever: 'idontcare'
        }
      }), err => {
        assume(err.statusCode).equals(404);
        assume(err.message).includes('Nock: No match for request');
        // Verify it retries
        assume(err['attempt#0']).is.truthy();
        assume(err['attempt#2']).is.truthy();

        nock.cleanAll();
        next();
      });
    });

    it('can also send stringified data', function (next) {
      next = assume.wait(2, next);

      nock(uri)
        .post('/v2/build')
        .reply(200, function reply(uri, body) {
          assume(body.data.name).equals('foo-bar');
          assume(body.promote).equals(true);
          nock.cleanAll();
          next();

          return {};
        });

      carpenter.build({
        data: JSON.stringify(options.data)
      }, next);
    });

    it('sends a request to /build when using v1 API', function (next) {
      carpenter = new Carpenter({
        uri,
        version: 'v1'
      });

      next = assume.wait(2, next);

      nock(uri)
        .post('/build')
        .reply(200, function reply(uri, body) {
          assume(body.name).equals('foo-bar');
          nock.cleanAll();
          next();

          return {};
        });

      carpenter.build({
        data: { name: 'foo-bar' }
      }, next);
    });
  });

  describe('#cancel', function () {
    var options = {
      pkg: 'foo-bar',
      version: '1.0.0',
      env: 'prod'
    };

    it('sends a request to /cancel', function (next) {
      next = assume.wait(2, next);

      nock(uri)
        .get('/cancel/foo-bar/1.0.0/prod')
        .reply(200, function reply(uri, body) {
          assume(uri).equals('/cancel/foo-bar/1.0.0/prod');
          assume(body).to.be.falsey();
          nock.cleanAll();
          next();

          return {};
        });

      carpenter.cancel(options, next);
    });

    it('sends a request to /cancel with retries and fails after retries', next => {
      carpenter = new Carpenter({
        url: uri,
        retry: {
          retries: 3,
          min: 100,
          max: 500
        }
      });
      nock(uri, {
        badheaders: ['whatever']
      })
        .post('/cancel/foo-bar/1.0.0/prod')
        .reply(200);

      carpenter.build(Object.assign({}, options, {
        headers: {
          whatever: 'idontcare'
        }
      }), err => {
        assume(err.statusCode).equals(404);
        assume(err.message).includes('Nock: No match for request');
        // Verify it retries
        assume(err['attempt#0']).is.truthy();
        assume(err['attempt#2']).is.truthy();

        nock.cleanAll();
        next();
      });
    });

    it('has optional env parameter', function (next) {
      next = assume.wait(2, next);

      delete options.env;

      nock(uri)
        .get('/cancel/foo-bar/1.0.0')
        .reply(200, function reply(uri, body) {
          assume(uri).equals('/cancel/foo-bar/1.0.0');
          assume(body).to.be.falsey();
          nock.cleanAll();
          next();

          return {};
        });

      carpenter.cancel(options, next);
    });
  });
});


index.js
'use strict';

var request = require('hyperquest');
var url = require('url');
var Retryme = require('retryme');

//
// Methods that require an `application/json` header.
//
var methods = ['POST', 'PUT'];

/**
 * Carpenter API client.
 *
 * @constructor
 * @param {Object|String} opts Options for root URL of carpenter service
 * @param {String} opts.url The root URL of the carpenter service
 * @param {String} opts.uri The root URL of the carpenter service
 * @param {String} opts.href The href for root URL of the carpenter service
 * @param {String} opts.protocol Protocol for root URL of the carpenter service
 * @param {String} opts.version The carpenter build API version to use (defaults to v2)
 * @public
 */
function Carpenter(opts) {
  // eslint-disable-next-line no-new
  if (!this) new Carpenter(opts);

  if (typeof opts === 'string') {
    this.base = opts;
  } else if (opts.protocol && opts.href) {
    this.base = url.format(opts);
  } else if (opts.url || opts.uri) {
    this.base = opts.url || opts.uri;
  } else {
    throw new Error('Carpenter URL required');
  }

  this.version = (opts.version === '1' || opts.version === 'v1' || opts.version === 1) ? '' : 'v2';

  //
  // Handle all possible cases
  //
  this.base = typeof this.base === 'object'
    ? url.format(this.base)
    : this.base;

  this.timeout = opts.timeout || 0;
  this.agent = opts.agent;
  this.retryOpts = opts.retry || { retries: 5, min: 500, max: 10000 };
}

/**
 * Trigger a new build.
 *
 * @param {Object} options Configuration.
 * @param {Function} next Completion callback.
 * @returns {Stream} the request
 * @private
 */
Carpenter.prototype.build = function build(options, next) {
  options = options || {};
  options.method = options.method || 'POST';

  return this.send([
    this.version,
    'build'
  ].filter(Boolean).join('/'), options, next);
};

/**
 * Cancel a build. The parameters are provided by options and filtered to allow
 * optional parameters.
 *
 * @param {Object} options Configuration.
 * @param {Function} next Completion callback.
 * @returns {Stream} the request
 * @private
 */
Carpenter.prototype.cancel = function cancel(options, next) {
  options = options || {};
  options.method = 'GET';

  return this.send([
    'cancel',
    options.pkg,
    options.version,
    options.env
  ].filter(Boolean).join('/'), options, next);
};

/**
 * Internal API for sending data.
 *
 * @param {String} pathname Pathname we need to hit.
 * @param {Object} options Hyperquest options
 * @param {Function} done Completion callback.
 * @returns {Stream} the request
 * @api private
 */
Carpenter.prototype.send = function send(pathname, options, done) {
  var base = url.parse(this.base),
    data = false,
    req;

  if (typeof pathname === 'object') {
    options = pathname;
    pathname = null;
  }

  if (typeof options === 'function') {
    done = options;
    options = {};
  }

  options.agent = this.agent;
  options.headers = options.headers || {};
  options.timeout = options.timeout || this.timeout;

  base.pathname = pathname || options.pathname || '/';

  //
  // Setup options from method and optional data.
  //
  data = options.data;
  if (typeof data === 'object' || ~methods.indexOf(options.method)) {
    options.method = options.method || 'POST';
    options.headers['Content-Type'] = 'application/json';
  }

  var operation = new Retryme(this.retryOpts);

  operation.attempt(next => {
    //
    // Setup hyperquest to formatted URL with retries.
    //
    req = request(url.format(base), options, next);

    //
    // Write JSON data to the request if provided. This is only required for
    // POST and PUT as hyperquest by default ends GET, DELETE and HEAD requests.
    //
    try {
      if (~methods.indexOf(options.method)) {
        req.end(typeof data === 'object' ? JSON.stringify(data) : data);
      }
    } catch (error) {
      return next(error);
    }
  }, done);


  return req;
};

//
// Expose the interface.
//
module.exports = Carpenter;
